scene/resources/cubemesh.cpp

```diff
@@ -0,0 +1,396 @@
+		for (i = 0; i <= subdivide_d; i++) {
+			float u = i;
+			float v = j;
+			u /= (3.0 * subdivide_d);
+			v /= (2.0 * subdivide_h);
+
+			/* right */
+			points.push_back(Vector3(-start_pos.x, -y, -z));
+			normals.push_back(Vector3(1.0, 0.0, 0.0));
+			ADD_TANGENT(0.0, 0.0, 1.0, -1.0);
+			uvs.push_back(Vector2(onethird + u, v));
+			point++;
+
+			/* left */
+			points.push_back(Vector3(start_pos.x, -y, z));
+			normals.push_back(Vector3(-1.0, 0.0, 0.0));
+			ADD_TANGENT(0.0, 0.0, -1.0, -1.0);
+			uvs.push_back(Vector2(u, 0.5 + v));
+			point++;
+
+			if (i > 0 && j > 0) {
+				int i2 = i * 2;
+
+				/* right */
+				indices.push_back(prevrow + i2 - 2);
+				indices.push_back(prevrow + i2);
+				indices.push_back(thisrow + i2 - 2);
+				indices.push_back(prevrow + i2);
+				indices.push_back(thisrow + i2);
+				indices.push_back(thisrow + i2 - 2);
+
+				/* left */
+				indices.push_back(prevrow + i2 - 1);
+				indices.push_back(prevrow + i2 + 1);
+				indices.push_back(thisrow + i2 - 1);
+				indices.push_back(prevrow + i2 + 1);
+				indices.push_back(thisrow + i2 + 1);
+				indices.push_back(thisrow + i2 - 1);
+			};
+
+			z += size.z / subdivide_d;
+		};
+
+		y += size.y / subdivide_h;
+		prevrow = thisrow;
+		thisrow = point;
+	};
+
+	/* top + bottom */
+	z = start_pos.z;
+	thisrow = point;
+	prevrow = 0;
+	for (j = 0; j <= subdivide_d; j++) {
+		x = start_pos.x;
+		for (i = 0; i <= subdivide_w; i++) {
+			float u = i;
+			float v = j;
+			u /= (3.0 * subdivide_w);
+			v /= (2.0 * subdivide_d);
+
+			/* top */
+			points.push_back(Vector3(-x, -start_pos.y, -z));
+			normals.push_back(Vector3(0.0, 1.0, 0.0));
+			ADD_TANGENT(1.0, 0.0, 0.0, -1.0);
+			uvs.push_back(Vector2(onethird + u, 0.5 + v));
+			point++;
+
+			/* bottom */
+			points.push_back(Vector3(x, start_pos.y, -z));
+			normals.push_back(Vector3(0.0, -1.0, 0.0));
+			ADD_TANGENT(-1.0, 0.0, 0.0, -1.0);
+			uvs.push_back(Vector2(twothirds + u, 0.5 + v));
+			point++;
+
+			if (i > 0 && j > 0) {
+				int i2 = i * 2;
+
+				/* top */
+				indices.push_back(prevrow + i2 - 2);
+				indices.push_back(prevrow + i2);
+				indices.push_back(thisrow + i2 - 2);
+				indices.push_back(prevrow + i2);
+				indices.push_back(thisrow + i2);
+				indices.push_back(thisrow + i2 - 2);
+
+				/* bottom */
+				indices.push_back(prevrow + i2 - 1);
+				indices.push_back(prevrow + i2 + 1);
+				indices.push_back(thisrow + i2 - 1);
+				indices.push_back(prevrow + i2 + 1);
+				indices.push_back(thisrow + i2 + 1);
+				indices.push_back(thisrow + i2 - 1);
+			};
+
+			x += size.x / subdivide_w;
+		};
+
+		z += size.z / subdivide_d;
+		prevrow = thisrow;
+		thisrow = point;
+	};
+
+	Array arr;
+	arr.resize(VS::ARRAY_MAX);
+	arr[VS::ARRAY_VERTEX] = points;
+	arr[VS::ARRAY_NORMAL] = normals;
+	arr[VS::ARRAY_TANGENT] = tangents;
+	arr[VS::ARRAY_TEX_UV] = uvs;
+	arr[VS::ARRAY_INDEX] = indices;
+
+	// out with the old
+	while (VisualServer::get_singleton()->mesh_get_surface_count(mesh) > 0) {
```

Comment:
Any request of info from VisualServer (in this case mesh_get_surface_count()) stalls the rendering thread, so it's better avoided